packages/block-editor/src/hooks/block-rename-ui.js

```diff
@@ -0,0 +1,234 @@
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent, useInstanceId } from '@wordpress/compose';
+import { addFilter } from '@wordpress/hooks';
+import { __, sprintf } from '@wordpress/i18n';
+import { getBlockSupport } from '@wordpress/blocks';
+import {
+	MenuItem,
+	__experimentalHStack as HStack,
+	__experimentalVStack as VStack,
+	Button,
+	TextControl,
+	Modal,
+} from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import { speak } from '@wordpress/a11y';
+
+/**
+ * Internal dependencies
+ */
+import {
+	BlockSettingsMenuControls,
+	useBlockDisplayInformation,
+	InspectorControls,
+} from '../components';
+
+const emptyString = ( testString ) => testString?.trim()?.length === 0;
+
+function RenameModal( { blockName, originalBlockName, onClose, onSave } ) {
+	const [ editedBlockName, setEditedBlockName ] = useState( blockName );
+
+	const nameHasChanged = editedBlockName !== blockName;
+	const nameIsOriginal = editedBlockName === originalBlockName;
+
+	const isNameValid =
+		( nameHasChanged || nameIsOriginal ) &&
+		! emptyString( editedBlockName );
+
+	const autoSelectInputText = ( event ) => event.target.select();
+
+	const dialogDescription = useInstanceId(
+		RenameModal,
+		`block-editor-rename-modal__description`
+	);
+
+	const handleSubmit = () => {
+		// Must be assertive to immediately announce change.
+		speak(
+			sprintf(
+				/* translators: %1$s: type of update (either reset of changed). %2$s: new name/label for the block */
+				__( 'Block name %1$s to: "%2$s".' ),
+				nameIsOriginal ? __( 'reset' ) : __( 'changed' ),
+				editedBlockName
+			),
+			'assertive'
+		);
+
+		onSave( editedBlockName );
+
+		// Immediate close avoids ability to hit save multiple times.
+		onClose();
+	};
+
+	return (
+		<Modal
+			title={ __( 'Rename block' ) }
+			onRequestClose={ onClose }
+			overlayClassName="block-editor-block-rename-modal"
+			aria={ {
+				describedby: dialogDescription,
+			} }
+		>
+			<p id={ dialogDescription }>
+				{ __( 'Choose a custom name for this block.' ) }
+			</p>
+			<form
+				onSubmit={ ( e ) => {
+					e.preventDefault();
+
+					if ( ! isNameValid ) {
+						return;
+					}
```

Comment:
The whole logic here is tricky.

I've updated the whole thing so that we never automatically set a `value` in the input on blur.

Instead we allow the field to be submitted empty. This allows us to remove `required` from the input.